### ISSUE TITLE
[BUG] AST join slows down significantly with a small left table

### ISSUE THREAD

**Issue by abellina** [rapidsai/cudf#9461]
We found an issue late in 21.10 (https://github.com/NVIDIA/spark-rapids/issues/3736) with the AST driven inner join where given a smaller left side table our performance would drop significantly with AST, as opposed to using an unconditional join + a separate filter. This issue is to discuss whether cuDF could detect this and adjust how it handles such a join.

If the larger table is on the left, AST is at least ~1.3x faster than join + filter approach. If the smaller table is on the left, AST can be 10x slower than the join + filter approach. We can work around this for inner joins specifically, since we can swap in the plugin the sides fairly easily (https://github.com/NVIDIA/spark-rapids/issues/3832), but this is hacky and seems like it is something cuDF should be able to handle, especially for all the other types of joins where we can't just swap sides like this.

The reason for the swap of tables in Spark, is that there is logic to build left or build right, depending on the byte sizes of the inputs. The spark-rapids plugin will split into smaller chunks the left side (if building right), or the right side (if building left), streaming these chunks against the side of the join that wasn't split (aka the build side). In this case we went from a build-left regime which was good for the AST since the left side was not split, to a build-right regime when the left side started to get split.

The reason why these kernels are so sensitive tot he left side, is because that's what is getting used to compute the grid size:
```
  detail::grid_1d config(left_table->num_rows(), DEFAULT_JOIN_BLOCK_SIZE);
```

And in our example (https://github.com/NVIDIA/spark-rapids/issues/3736) the left table could be hundreds of rows, where the right table could be 1M rows.

**Comment by vyasr** [comment]
In early prototypes of the AST implementation there was logic in the kernels to automatically swap the left and right tables so that the kernel configuration would always run over the left table. I removed this because that code path was swapping the tables without making the necessary associated modifications to the expression (e.g. `operation(GREATER, 0, 1)` needs to be converted to `operation(GREATER, 1, 0)`), leading to incorrect results. It should definitely be possible to do this in a safe way, though, so I'll look into the best way to accomplish that.

**Comment by GregoryKimball** [comment]
@vyasr should we close this?

**Comment by jlowe** [comment]
I don't think this has been solved for any case other than inner joins (which allows swapping the tables).

**Comment by vyasr** [comment]
@jlowe is correct, this has only been solved for inner joins. Solving it for other types of joins will be substantially more complicated and will likely require an entirely new algorithm, and the new algorithm will most likely have measurably larger overhead. Perhaps there is a generic solution that does not introduce any more work, but I was not able to come up with one at the time that this issue was originally raised, even if I allowed for modification of the expression IIRC.

Most likely what we would need to do is some careful benchmarking of the current approach against a new approach for different data sizes to determine tradeoffs and perform a runtime dispatch to the appropriate code path depending on the size imbalance. That will also balloon compile time so it's something that we will have to consider very carefully.

**Comment by vyasr** [comment]
One possibly reasonable naive solution for might be to do a two pass approach that is an inner join followed by a left join with a suitable predicate to determine which rows had no matches in the inner join. That may still be sufficient to overcome the current performance bottlenecks in some cases without doubling the number of kernels or requiring the development of entirely new ones.
